Add options to change name and nextRun in retryJob

mea36 opened the pull request:
Giving this to Chirag to complete over the next few days since I'll be out.

I hate that the order of parameters for `RetryableException` doesn't make any sense, but I don't want to change it so we can be backwards compatible.  
We can create another exception type so the that order of parameters makes more sense, but that doesn't seem worth it.

I didn't update the package version yet, I'll wait for the review before doing that

Closes https://github.com/Expensify/Expensify/issues/59735

## Tests
- Create a job, retry it with changing it's name, confirm the name was changed
```
>>> use Expensify\Bedrock\Client;use Expensify\Bedrock\DB;use Expensify\Bedrock\Jobs;$c = new Client(); $jobs = new Jobs($c);$db = new DB($c)
=> Expensify\Bedrock\DB {#395}
>>> $jobs->createJob('foo')
=> [
     "codeLine" => "200 OK",
     "code" => 200,
     "headers" => [
       "commitCount" => "10312",
       "nodeName" => "bedrock",
       "peekTime" => "70",
       "processTime" => "9770",
       "totalTime" => "26502",
       "Content-Length" => "14",
     ],
     "rawBody" => "{"jobID":2440}",
     "body" => [
       "jobID" => 2440,
     ],
   ]
>>> $j = $jobs->getJob('foo')
=> [
     "codeLine" => "200 OK",
     "code" => 200,
     "headers" => [
       "commitCount" => "10313",
       "nodeName" => "bedrock",
       "peekTime" => "4558",
       "processTime" => "2143",
       "totalTime" => "15718",
       "Content-Length" => "37",
     ],
     "rawBody" => "{"data":[],"jobID":2439,"name":"foo"}",
     "body" => [
       "data" => [],
       "jobID" => 2439,
       "name" => "foo",
     ],
   ]
>>> $db->query("select * from jobs where jobID = ".$j['body']['jobID'])->toArray()['rawBody']
=> "{"headers":["created","jobID","state","name","nextRun","lastRun","repeat","data","priority","parentJobID","retryAfter"],"rows":[["2017-09-19 21:46:18",2439,"RUNNING","foo","2017-09-19 21:49:19","2017-09-19 22:09:16","",[],500,0,""]]}"
>>> $jobs->retryJob($j['body']['jobID'], 0, $j['body']['data'], 'foo?bar')
=> [
     "codeLine" => "200 OK",
     "code" => 200,
     "headers" => [
       "commitCount" => "10314",
       "nodeName" => "bedrock",
       "peekTime" => "55",
       "processTime" => "2168",
       "totalTime" => "6109",
       "Content-Length" => "0",
     ],
     "rawBody" => "",
     "body" => [],
   ]
>>> $db->query("select * from jobs where jobID = ".$j['body']['jobID'])->toArray()['rawBody']
=> "{"headers":["created","jobID","state","name","nextRun","lastRun","repeat","data","priority","parentJobID","retryAfter"],"rows":[["2017-09-19 21:46:18",2439,"QUEUED","foo?bar","2017-09-19 22:09:54","2017-09-19 22:09:16","",[],500,0,""]]}"
```
- Get the previous job, retry it with changing it's name and the nextRun time, confirm the name and nextRun were changed
```
>>> $j = $jobs->getJob('foo?bar')
=> [
     "codeLine" => "200 OK",
     "code" => 200,
     "headers" => [
       "commitCount" => "10369",
       "nodeName" => "bedrock",
       "peekTime" => "2719",
       "processTime" => "1173",
       "totalTime" => "9072",
       "Content-Length" => "41",
     ],
     "rawBody" => "{"data":[],"jobID":2444,"name":"foo?bar"}",
     "body" => [
       "data" => [],
       "jobID" => 2444,
       "name" => "foo?bar",
     ],
   ]
>>> $db->query("select * from jobs where jobID = ".$j['body']['jobID'])->toArray()['rawBody']
=> "{"headers":["created","jobID","state","name","nextRun","lastRun","repeat","data","priority","parentJobID","retryAfter"],"rows":[["2017-09-19 22:08:04",2440,"RUNNING","foo","2017-09-19 22:08:04","2017-09-19 23:29:38","",{},500,0,""]]}"
>>> $jobs->retryJob($j['body']['jobID'], 0, $j['body']['data'], 'foo?bar=baz', '2017-09-23 22:08:04')
=> [
     "codeLine" => "200 OK",
     "code" => 200,
     "headers" => [
       "commitCount" => "10317",
       "nodeName" => "bedrock",
       "peekTime" => "44",
       "processTime" => "1912",
       "totalTime" => "10998",
       "Content-Length" => "0",
     ],
     "rawBody" => "",
     "body" => [],
   ]
>>> $db->query("select * from jobs where jobID = ".$j['body']['jobID'])->toArray()['rawBody']
=> "{"headers":["created","jobID","state","name","nextRun","lastRun","repeat","data","priority","parentJobID","retryAfter"],"rows":[["2017-09-19 22:08:04",2440,"QUEUED","foo?bar=baz","2017-09-23 22:08:04","2017-09-19 23:29:38","",[],500,0,""]]}"
```
- Create a job, throw a RetryableException, confirm that the name and nextRun time are changed
```
>>> use Expensify\Bedrock\Exceptions\Jobs\RetryableException;
>>> $jobs->createJob('baz')
=> [
     "codeLine" => "200 OK",
     "code" => 200,
     "headers" => [
       "commitCount" => "10318",
       "nodeName" => "bedrock",
       "peekTime" => "61",
       "processTime" => "1430",
       "totalTime" => "5477",
       "Content-Length" => "14",
     ],
     "rawBody" => "{"jobID":2442}",
     "body" => [
       "jobID" => 2442,
     ],
   ]
>>> $j = $jobs->getJob('baz')
=> [
     "codeLine" => "200 OK",
     "code" => 200,
     "headers" => [
       "commitCount" => "10319",
       "nodeName" => "bedrock",
       "peekTime" => "3783",
       "processTime" => "2229",
       "totalTime" => "10588",
       "Content-Length" => "37",
     ],
     "rawBody" => "{"data":{},"jobID":2442,"name":"baz"}",
     "body" => [
       "data" => [],
       "jobID" => 2442,
       "name" => "baz",
     ],
   ]
>>> $db->query("select * from jobs where jobID = ".$j['body']['jobID'])->toArray()['rawBody']
=> "{"headers":["created","jobID","state","name","nextRun","lastRun","repeat","data","priority","parentJobID","retryAfter"],"rows":[["2017-09-19 23:33:19",2442,"RUNNING","baz","2017-09-19 23:33:19","2017-09-19 23:33:28","",{},500,0,""]]}"
>>> $e = new RetryableException('retrying this jawn', 0, null, null, 'baz?bar', '2017-09-23 22:08:04')
=> Expensify\Bedrock\Exceptions\Jobs\RetryableException {#386
     #message: "retrying this jawn",
     #code: 666,
     #file: "/home/vagrant/.config/composer/vendor/psy/psysh/src/Psy/ExecutionLoop/Loop.php(90) : eval()'d code",
     #line: 1,
   }
>>> $jobs->retryJob($j['body']['jobID'], $e->getDelay(), $j['body']['data'], $e->getName(), $e->getNextRun())
=> [
     "codeLine" => "200 OK",
     "code" => 200,
     "headers" => [
       "commitCount" => "10323",
       "nodeName" => "bedrock",
       "peekTime" => "47",
       "processTime" => "5983",
       "totalTime" => "14693",
       "Content-Length" => "0",
     ],
     "rawBody" => "",
     "body" => [],
   ]
>>> $db->query("select * from jobs where jobID = ".$j['body']['jobID'])->toArray()['rawBody']
=> "{"headers":["created","jobID","state","name","nextRun","lastRun","repeat","data","priority","parentJobID","retryAfter"],"rows":[["2017-09-19 23:37:37",2443,"QUEUED","baz?bar","2017-09-23 22:08:04","2017-09-19 23:37:46","",[],500,0,""]]}"
```
- Create a job, test that `RetryableException` with `delay` still works
```
>>> $jobs->createJob('bar')
=> [
     "codeLine" => "200 OK",
     "code" => 200,
     "headers" => [
       "commitCount" => "10374",
       "nodeName" => "bedrock",
       "peekTime" => "59",
       "processTime" => "792",
       "totalTime" => "8992",
       "Content-Length" => "14",
     ],
     "rawBody" => "{"jobID":2446}",
     "body" => [
       "jobID" => 2446,
     ],
   ]
>>> $j = $jobs->getJob('bar')
=> [
     "codeLine" => "200 OK",
     "code" => 200,
     "headers" => [
       "commitCount" => "10375",
       "nodeName" => "bedrock",
       "peekTime" => "3390",
       "processTime" => "1103",
       "totalTime" => "10461",
       "Content-Length" => "37",
     ],
     "rawBody" => "{"data":{},"jobID":2446,"name":"bar"}",
     "body" => [
       "data" => [],
       "jobID" => 2446,
       "name" => "bar",
     ],
   ]
>>> $db->query("select * from jobs where jobID = ".$j['body']['jobID'])->toArray()['rawBody']
=> "{"headers":["created","jobID","state","name","nextRun","lastRun","repeat","data","priority","parentJobID","retryAfter"],"rows":[["2017-09-21 19:15:29",2446,"RUNNING","bar","2017-09-21 19:15:29","2017-09-21 19:15:35","",{},500,0,""]]}"
>>> $e = new RetryableException('retrying this jawn', 600)
=> Expensify\Bedrock\Exceptions\Jobs\RetryableException {#405
     #message: "retrying this jawn",
     #code: 666,
     #file: "/home/vagrant/.config/composer/vendor/psy/psysh/src/Psy/ExecutionLoop/Loop.php(90) : eval()'d code",
     #line: 1,
   }
>>> $jobs->retryJob($j['body']['jobID'], $e->getDelay(), $j['body']['data'], $e->getName(), $e->getNextRun())
=> [
     "codeLine" => "200 OK",
     "code" => 200,
     "headers" => [
       "commitCount" => "10376",
       "nodeName" => "bedrock",
       "peekTime" => "43",
       "processTime" => "1354",
       "totalTime" => "5470",
       "Content-Length" => "0",
     ],
     "rawBody" => "",
     "body" => [],
   ]
>>> $db->query("select * from jobs where jobID = ".$j['body']['jobID'])->toArray()['rawBody']
=> "{"headers":["created","jobID","state","name","nextRun","lastRun","repeat","data","priority","parentJobID","retryAfter"],"rows":[["2017-09-21 19:15:29",2446,"QUEUED","bar","2017-09-21 19:26:24","2017-09-21 19:15:35","",[],500,0,""]]}"
```

iwiznia commented on src/Jobs.php:
```diff
@@ -346,20 +346,24 @@ public function failJob($jobID)
      * @return array
      */
-    public function retryJob($jobID, $delay = 0, $data = [])
+    public function retryJob($jobID, $delay = 0, $data = null, $name = '', $nextRun = '')
```
Add type annotations for the new params

iwiznia commented on bin/BedrockWorkerManager.php:
```diff
@@ -264,7 +264,7 @@
                                 'extraParams' => $extraParams,
                             ]);
-                            $jobs->retryJob($job['jobID'], $e->getDelay(), $worker->getData());
+                            $jobs->retryJob($job['jobID'], $e->getDelay(), $worker->getData(), $e->getName(), $e->getNextRun());
```
Curious to know why you decided to add this to the RetryableException instead of to do worker. Like with the data, we set it in the workers via `setParam` and then get it here via `getData`. I think name and nextRun should be there too. 
Also, now you have nextRun and delay, which are exclusive. Which one takes precedence? Should we remove the support for delay and make it use `nextRun` internally?

mea36 replied:
> Curious to know why you decided to add this to the RetryableException instead of to do worker

I don't understand your question.  We need to add it to the exception because the jobs will throw `RetryableException`s with these values

> I think name and nextRun should be there too

So you think the job should know what it's `nextRun` value should be? That doesn't make sense to me.  Should the job know when it's `delay` value should be as well?

> Also, now you have nextRun and delay, which are exclusive. Which one takes precedence?

https://github.com/Expensify/Bedrock/blob/358f98e44a360012ac894e821c76342c6bba3e2a/plugins/Jobs.cpp#L812-L816

> Should we remove the support for delay and make it use `nextRun` internally?

I'm not comfortable doing this because this is an open source library and we don't have a versioning system set in place yet with a changelog to convey what we're changing.  If we change things to suit our purpose, we're potentially breaking code for other people

iwiznia replied:
> I don't understand your question. We need to add it to the exception because the jobs will throw RetryableExceptions with these values

I know, but the Worker instance still exists and we can data form it. Again, look at how we update the data of the jobs when finishing/retrying them.

> So you think the job should know what it's nextRun value should be? That doesn't make sense to me. Should the job know when it's delay value should be as well?

Yes, that's exactly what I'm saying and in hindsight yes, it should know it's delay too (but we don't have to change this now).

> I'm not comfortable doing this because this is an open source library and we don't have a versioning system set in place yet with a changelog to convey what we're changing. If we change things to suit our purpose, we're potentially breaking code for other people

1. I said internally, so users would still throw a RetryableException, but internally it would just set the nextRun.
2. We do have versioning in place, this is a composer lib, so it has versioning. If you think someone could be using this and could break their app, then just increment the major part of the version number. 
3. I definitely agree we need a readme.

mea36 replied:
> I know, but the Worker instance still exists and we can data form it. Again, look at how we update the data of the jobs when finishing/retrying them. 

I think that this is odd.  It doesn't make sense that the worker should know it's `nextRun` time, but it does make sense that the worker should know that the job's `data` is.  The worker should just take the job data and script name and execute the script with the data.  

> I said internally, so users would still throw a RetryableException, but internally it would just set the nextRun.

So the `Bedrock-PHP` layer will handle the logic for figuring out the `nextRun`?  I don't agree with that.  Sorry if I'm missing something obvious here

fnwbr commented on src/Exceptions/Jobs/RetryableException.php:
```diff
@@ -38,4 +53,24 @@ public function getDelay()
+     *
+     * @return string
+     */
+    public function getName()
```
Missing return type annotation

fnwbr commented on src/Exceptions/Jobs/RetryableException.php:
```diff
@@ -38,4 +53,24 @@ public function getDelay()
+     *
+     * @return string
+     */
+    public function getNextRun()
```
Missing return type annotation

iwiznia commented:
When you are ready to create the version for this, watch out since I created 1.0.33 for https://github.com/Expensify/Bedrock-PHP/pull/41

mea36 commented:
I found a bug and fixed it https://github.com/Expensify/Bedrock-PHP/pull/40/commits/007ca0a488fedf3140320e76e974a155c7862848
I also updated the tests in the description to test for this - using RetryableException without `name` or `nextRun` parameters
